agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,821 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/connect/ca"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/logging"
+	"github.com/hashicorp/go-hclog"
+	uuid "github.com/hashicorp/go-uuid"
+)
+
+type CAState string
+
+const (
+	CAStateUninitialized     CAState = "UNINITIALIZED"
+	CAStateInitializing              = "INITIALIZING"
+	CAStateReady                     = "READY"
+	CAStateRenewIntermediate         = "RENEWING"
+	CAStateReconfig                  = "RECONFIGURING"
+)
+
+// CAManager is a wrapper around CA operations such as updating roots, an intermediate
+// or the configuration. All operations should go through the CAManager in order to
+// avoid data races.
+type CAManager struct {
+	srv    *Server
+	logger hclog.Logger
+
+	// provider is the current CA provider in use for Connect. This is
+	// only non-nil when we are the leader.
+	provider ca.Provider
+
+	// providerRoot is the CARoot that was stored along with the ca.Provider
+	// active. It's only updated in lock-step with the provider. This prevents
+	// races between state updates to active roots and the fetch of the provider
+	// instance.
+	providerRoot *structs.CARoot
+	providerLock sync.RWMutex
+
+	// primaryRoots is the most recently seen state of the root CAs from the primary datacenter.
+	// This is protected by the stateLock and updated by initializeCA and the root CA watch routine.
+	primaryRoots structs.IndexedCARoots
+
+	// actingSecondaryCA is whether this datacenter has been initialized as a secondary CA.
+	actingSecondaryCA bool
+	state             CAState
+	stateLock         sync.RWMutex
+}
+
+func NewCAManager(srv *Server) *CAManager {
+	return &CAManager{
+		srv:    srv,
+		logger: srv.loggers.Named(logging.Connect),
+		state:  CAStateUninitialized,
+	}
+}
+
+// setState attempts to update the CA state to the given state.
+// If the current state is not READY, this will fail. The only exception is when
+// the current state is UNINITIALIZED, and the function is called with CAStateInitializing.
+func (c *CAManager) setState(newState CAState) error {
+	c.stateLock.RLock()
+	state := c.state
+	c.stateLock.RUnlock()
+
```

Comment:
You have a minor bug here with the double checked locking implementation. Usually you would do a cheap lock (the read lock) first to short circuit doing the expensive lock (write lock) to do the real work. In that case the first check is purely a perf optimization so the rest has to function correctly even if the first block were erased. But in this implementation you don't re-check the conditionals from line 73 again with the write lock held, since the state could've drifted in the intervening clock ticks.

Practically speaking would it be easier here to change this to be:
```
c.stateLock.Lock()
defer c.stateLock.Unlock()
// your if conditional, error returns, and state update code
```
Usually the read/write lock split makes sense only if you expect a high degree of "no-op" writes, but given the body of this function that seems less likely